hopsworks-api/src/main/java/io/hops/hopsworks/api/jupyter/JupyterService.java

```diff
@@ -415,7 +430,30 @@ private void stop(String hdfsUser) throws AppException {
       throw new AppException(Response.Status.BAD_REQUEST.getStatusCode(),
           "Could not find Jupyter entry for user: " + hdfsUser);
     }
+
+    Users user = userFacade.findByEmail(loggedinemail);
+
+    List<LivyMsg.Session> sessions = livyService.
+        getLivySessionsForProjectUser(project, user, ProjectServiceEnum.JUPYTER);
+
     livyService.deleteAllLivySessions(hdfsUser, ProjectServiceEnum.JUPYTER);
+
+    int retries = 3;
+    while(retries > 0) {
+      if (livyService.getLivySessionsForProjectUser(project, user, ProjectServiceEnum.JUPYTER).size() > 0) {
+        LOGGER.log(Level.SEVERE, "Failed previous attempt to delete livy sessions for project " + project.getName() +
+            " user " + hdfsUser + ", retrying...");
+        livyService.deleteAllLivySessions(hdfsUser, ProjectServiceEnum.JUPYTER);
+      } else {
+        break;
```

Comment:
@robzor92 what's the purpose of this `else` clause?

Reply:
It will try 3 times to kill the Livy sessions for the projectuser. I observed that sometimes they were not shutdown and the experiment (YARN job) would keep running so I added a mechanism to make the shutdown more reliable. The else clause is used to break whenever there are no running livy sessions during the 3 retries.

Reply:
This happens extremely rarely, like 1/50 times and I haven't been able to locate the exact cause therefore this mechanism should remedy the problem.

Reply:
Wouldn't this be a better alternative then? 
`while(retries > 0 && livyService.getLivySessionsForProjectUser(project, user, ProjectServiceEnum.JUPYTER).size() > 0) ` without the `else`.

Reply:
Sure changing it